guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+🚫
+
+``` typescript
+describe('KeyringController', () => {
+  it('adds a new account to the given keyring', () => {
+    // ...
+  });
+
+  it('removes the account identified by the given address from its associated keyring', () => {
+    // ...
+  });
+});
+```
+
+✅
+
+``` typescript
+describe('KeyringController', () => {
+  describe('addAccount', () => {
+    it('adds a new account to the given keyring', () => {
+      // ...
+    });
+  });
+
+  describe('removeAccount', () => {
+    it('removes the account identified by the given address from its associated keyring', () => {
+      // ...
+    });
+  });
+});
+```
+
+## 💡 Use `describe` to group tests for scenarios
+
+If you have many tests that verify the behavior of a piece of code under a particular condition, you might find it helpful to wrap those tests with a `describe` block. Use a phrase starting with `if ...` or `when ...` so as to form a sentence when combined with the test name.
+
+1️⃣
+
+``` typescript
+it('calls addToken on the tokens controller if "addToken" is true', () => {
+  // ...
+})
+
+it('adds the new token to the state if "addToken" is true', () => {
+  // ...
+})
+
+it('returns a promise that resolves to true if "addToken" is true', () => {
+  // ...
+})
+```
+
+2️⃣
+
+``` typescript
+describe('if "addToken" is true', () => {
+  it('calls addToken on the tokens controller', () => {
+    // ...
+  });
+
+  it('adds the new token to the state', () => {
+    // ...
+  });
+
+  it('returns a promise that resolves to true', () => {
+    // ...
+  });
+});
+```
+
+## Use `it` to specify known behavior for a unit of code
+
+Use `it` to describe the behavior of a piece of code from the consumer's perspective as it stands today, not as it ought to be.
+
+Do not use "should" at the start of the test name. The official Jest documentation [omits this word from their examples](https://jestjs.io/docs/next/getting-started), and it creates noise when reviewing the list of tests that Jest outputs after a run.
+
+Do not repeat the name of the function or method in the name of the test if it is possible to communicate what a consumer expects more clearly.
+
+### Examples
+
+🚫
+
+``` typescript
+it("should not stop the block tracker", () => {
+  // ...
+});
+```
+
+✅
+
+``` typescript
+it("does not stop the block tracker", () => {
+  // ...
+});
+```
+
+🚫
+
+``` typescript
+describe('TokensController', () => {
+  it("addToken", () => {
+    // ...
+  });
+});
+```
+
+🚫
+
+``` typescript
+describe('TokensController', () => {
+  it('adds a token', () => {
+    // ...
+  });
+})
+```
+
+✅
+
+``` typescript
+describe('TokensController', () => {
+  describe('addToken', () => {
+    it('adds the given token to the "tokens" array in state', () => {
```

Comment:
Nit: "tokens array" here seems like an implementaiton detail. Ideally the test descriptions would describe the behavior in high-level terms, so it's clear to the reader why we want this to happen, and so that we don't have to update the description if we refactor things (e.g. if we change the tokens array to a set, for example).
```suggestion
    it('adds the given token to state', () => {
```

Reply:
Good point. Updated in 66469342083e36e0bc7b9e0bbecf9b15b65de259...9f892512f284e55f56defb2bdd136f675d546710.